### PR TITLE
354 pagination better cache

### LINKED ISSUE #633: Tab nav disappears when navigating to a site with no records

### PR DESCRIPTION
**Pull request by jlhogan**
No pagination yet (lol) but SWR has built in things for fetching and caching pages of data, so it will get done today OR ELSE

the tldr here is now swr is the cache, and you can mutate a list of cache things with a single item by using the provided mutate function if you've fetched the same key. Had to change the apiGet signature to take accessToken as an inline argument to avoid unnecessary code writing but if anyone has strong objection to that, let's talk.  we can tweak things like deduping interval or refetch interval or retry interval. the docs are actually v great: https://github.com/vercel/swr

also closes #633

### CODE REVIEW

File: client/src/containers/CreateRecord/CreateRecord.tsx [hunk 1]
```diff
@@ -104,18 +116,19 @@ const CreateRecord: React.FC = () => {
...
         const currentStepStatus = steps[indexOfCurrentStep].status({
           child: updatedChild,
         } as RecordFormProps);
 
-        if (currentStepStatus === 'complete') {
+        if (
+          currentStepStatus === 'complete' ||
+          currentStepStatus === 'exempt'
+        ) {
+          console.log('moving next step');
```
Comment (mmazanec22) [inline client/src/containers/CreateRecord/CreateRecord.tsx:131]: Console log

File: client/src/containers/Roster/RosterButtonsTable.tsx [hunk 2]
```diff
@@ -1,19 +1,17 @@
 import { Button, Calendar, TextWithIcon } from '@ctoec/component-library';
 import { Moment } from 'moment';
-import React, { useState } from 'react';
+import React, { useState, useContext } from 'react';
```
Comment (mmazanec22) [inline client/src/containers/Roster/RosterButtonsTable.tsx:3]: Looks like some of these imports are unused

File: client/src/containers/Roster/rosterUtils.tsx [hunk 3]
```diff
@@ -1,153 +1,68 @@
...
-  };
-}
-
+/**
+ * Does client-side data filtering, by site and/or month
+ * @param allChildren
+ * @param site
+ * @param month
+ */
 export function getFilteredChildren(
```
Comment (mmazanec22) [inline client/src/containers/Roster/rosterUtils.tsx:33]: A note here about how this is only for front end filtering, or maybe a func renaming, could make this clearer

File: src/controllers/reportingPeriods.ts [hunk 4]
```diff
@@ -1,12 +1,18 @@
...
 import moment from 'moment';
+import { FundingSource } from '../../client/src/shared/models';
 
 /**
  * Get all reporting periods, less than 3 years old
  * and less than 3 years in the future
  */
-export const getReportingPeriods = async () => {
+export const getReportingPeriods = async (shortFundingSource?: string) => {
+  console.log('fundingSource', shortFundingSource);
```
Comment (mmazanec22) [inline src/controllers/reportingPeriods.ts:11]: Console log

File: src/controllers/reportingPeriods.ts [hunk 5]
```diff
@@ -1,12 +1,18 @@
...
+  const fundingSource = Object.values(FundingSource).find((fs) =>
+    fs.startsWith(shortFundingSource)
+  );
+  console.log('looke dup', fundingSource);
```
Comment (mmazanec22) [inline src/controllers/reportingPeriods.ts:15]: Looke dup!

**Review by mmazanec22 — APPROVED** [review]

File: src/controllers/reportingPeriods.ts (hunk 5 above)
Comment (jlhogan) [reply src/controllers/reportingPeriods.ts:15]: :)

File: client/src/components/Forms/Enrollment/Funding/Fields/ContractSpace/ContractSpace.tsx [hunk 6]
```diff
@@ -10,15 +10,17 @@ import {
...
 import { SingleContractSpaceField } from './SingleContractSpace';
 import {
   ChangeFunding,
   ChangeEnrollment,
 } from '../../../../../../shared/payloads';
 import { fundingSpaceFormatter } from '../../../../../../utils/formatters';
-import DataCacheContext from '../../../../../../contexts/DataCacheContext/DataCacheContext';
 import { getValidationStatusForField } from '../../../../../../utils/getValidationStatus';
+import { useAuthenticatedSWR } from '../../../../../../hooks/useAuthenticatedSWR';
+import { stringify } from 'querystring';
```
[the remark this answers is not in the document]
Comment (jlhogan) [reply client/src/components/Forms/Enrollment/Funding/Fields/ContractSpace/ContractSpace.tsx:23]: i added `query-string` a while ago, and i think it's a bit better, however if `query-string` is in there and we can't remove it then maybe we stick with that? At least requires parsing the `?` and `#` off of search/hash, which is annoying. will TAL.

**Review by bamader — APPROVED** [review]
Looks good! Will probably need to read the docs myself before doing any coding or manipulating the cache but you're right, they do look pretty friendly.

**Pull request merged**